File: src/ios/CDVBMSAuthorizationManager.swift
/*
*     Copyright 2016 IBM Corp.
*     Licensed under the Apache License, Version 2.0 (the "License");
*     you may not use this file except in compliance with the License.
*     You may obtain a copy of the License at
*     http://www.apache.org/licenses/LICENSE-2.0
*     Unless required by applicable law or agreed to in writing, software
*     distributed under the License is distributed on an "AS IS" BASIS,
*     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*     See the License for the specific language governing permissions and
*     limitations under the License.
*/

import Foundation
import BMSCore
import BMSSecurity


@objc(CDVBMSAuthenticationManager) class CDVBMSAuthorizationManager : CDVPlugin {

    func initialize(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {

                guard let tenantId = command.arguments[0] as? String else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: CustomErrorMessages.invalidTenantId)
                    // call failure callback
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    return
                }

                let authManager = MCAAuthorizationManager.sharedInstance
                authManager.initialize(tenantId: tenantId)
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            });
        #else
            self.commandDelegate!.runInBackground({

                guard let tenantId = command.arguments[0] as? String else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: CustomErrorMessages.invalidTenantId)
                    // call failure callback
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    return
                }

                let authManager = MCAAuthorizationManager.sharedInstance
                authManager.initialize(tenantId: tenantId)
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            });

        #endif
    }

    func obtainAuthorizationHeader(_ command: CDVInvokedUrlCommand) {
        let authManager = MCAAuthorizationManager.sharedInstance;

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                authManager.obtainAuthorization{ (response: Response?, error: Error?) -> Void in
                    var responseString: String?

                    do {
                        if (error != nil) {
                            // process the error
                            try responseString = Utils.packResponse(response, error: error as NSError?)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: responseString)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        } else {
                            // process success
                            try responseString = Utils.packResponse(response)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: responseString)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        }
                    } catch {
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: CustomErrorMessages.errorParsingJSONResponse)
                        self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    }
                }


            });
        #else
            self.commandDelegate!.runInBackground({

                authManager.obtainAuthorization { (response: Response?, error: NSError?) -> Void in
                    var responseString: String?

                    do {
                        if (error != nil) {
                            // process the error
                            try responseString = Utils.packResponse(response, error: error)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        } else {
                            // process success
                            try responseString = Utils.packResponse(response)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        }
                    } catch {
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: CustomErrorMessages.errorParsingJSONResponse)
                        self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    }
                }

            });
        #endif

    }

    func isAuthorizationRequired(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {

                do {
                    let authManager = MCAAuthorizationManager.sharedInstance
                    let params = try self.unpackIsAuthorizationRequiredParams(command);

                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:authManager.isAuthorizationRequired(for: Int(params.statusCode), httpResponseAuthorizationHeader: params.authorizationHeaderValue))

                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                } catch {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid parameters passed to isAuthorizationRequired method")
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                }
            })
        #else
            self.commandDelegate!.runInBackground({

                do {
                    let authManager = MCAAuthorizationManager.sharedInstance
                    let params = try self.unpackIsAuthorizationRequiredParams(command);

                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool:authManager.isAuthorizationRequired(for: Int(params.statusCode), httpResponseAuthorizationHeader: params.authorizationHeaderValue))

                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                } catch {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: "Invalid parameters passed to isAuthorizationRequired method")
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                }
            })
        #endif
    }

    func clearAuthorizationData(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                authManager.clearAuthorizationData()
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                authManager.clearAuthorizationData()
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func getCachedAuthorizationHeader(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance

                if let authHeader: String = authManager.cachedAuthorizationHeader {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:authHeader)
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                } else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: CustomErrorMessages.noCachedAuthorizationHeader)
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                }
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance

                if let authHeader: String = authManager.cachedAuthorizationHeader {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:authHeader)
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                } else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: CustomErrorMessages.noCachedAuthorizationHeader)
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                }
            })
        #endif
    }

    func getUserIdentity(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let userIdentity: String = try Utils.stringifyResponse(authManager.userIdentity as AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:userIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs:CustomErrorMessages.errorObtainUserIdentity)
                }

                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let userIdentity: String = try Utils.stringifyResponse(authManager.userIdentity as! AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:userIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:CustomErrorMessages.errorObtainUserIdentity)
                }

                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func getAppIdentity(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let appIdentity: String = try Utils.stringifyResponse(authManager.appIdentity as AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:appIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs:CustomErrorMessages.errorObtainAppIdentity)
                }

                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let appIdentity: String = try Utils.stringifyResponse(authManager.appIdentity as! AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:appIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:CustomErrorMessages.errorObtainAppIdentity)
                }

                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func getDeviceIdentity(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let deviceIdentity: String = try Utils.stringifyResponse(authManager.deviceIdentity as AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:deviceIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs:CustomErrorMessages.errorObtainDeviceIdentity)
                }

                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                var pluginResult: CDVPluginResult? = nil

                do {
                    let deviceIdentity: String = try Utils.stringifyResponse(authManager.deviceIdentity as! AnyObject)
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:deviceIdentity)
                } catch {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:CustomErrorMessages.errorObtainDeviceIdentity)
                }

                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func getAuthorizationPersistencePolicy(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                let policy: PersistencePolicy = authManager.authorizationPersistencePolicy()
                var pluginResult: CDVPluginResult? = nil

                switch policy {
                    case PersistencePolicy.always:
                        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:PersistencePolicy.always.rawValue)
                    case PersistencePolicy.never:
                        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:PersistencePolicy.never.rawValue)
                    default:
                        pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs:CustomErrorMessages.invalidPolicyType)
                }

                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                let policy: PersistencePolicy = authManager.authorizationPersistencePolicy()
                var pluginResult: CDVPluginResult? = nil

                switch policy {
                case PersistencePolicy.always:
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:PersistencePolicy.always.rawValue)
                case PersistencePolicy.never:
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString:PersistencePolicy.never.rawValue)
                default:
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:CustomErrorMessages.invalidPolicyType)
                }

                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func setAuthorizationPersistencePolicy(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                guard let policy: String = command.arguments[0] as? String else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs:CustomErrorMessages.invalidPolicySpecified)
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    return
                }

                switch policy {
                    case PersistencePolicy.always.rawValue:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.always)
                    case PersistencePolicy.never.rawValue:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.never)
                    default:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.never)
                }

                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                guard let policy: String = command.arguments[0] as? String else {
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:CustomErrorMessages.invalidPolicySpecified)
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    return
                }

                switch policy {
                    case PersistencePolicy.always.rawValue:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.always)
                    case PersistencePolicy.never.rawValue:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.never)
                    default:
                        authManager.setAuthorizationPersistencePolicy(PersistencePolicy.never)
                }

                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif
    }

    func logout(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let authManager = MCAAuthorizationManager.sharedInstance
                    authManager.logout{ (response: Response?, error: Error?) -> Void in
                    var responseString: String?
                    do {
                        if (error != nil) {
                            // process the error
                            try responseString = Utils.packResponse(response, error: error as NSError?)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: responseString)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        } else {
                            // process success
                            try responseString = Utils.packResponse(response)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: responseString)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        }
                    } catch {
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: CustomErrorMessages.errorParsingJSONResponse)
                        self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    }
                }

            });
        #else
            self.commandDelegate!.runInBackground({
                let authManager = MCAAuthorizationManager.sharedInstance
                authManager.logout{ (response: Response?, error: NSError?) -> Void in
                    var responseString: String?
                    do {
                        if (error != nil) {
                            // process the error
                            try responseString = Utils.packResponse(response, error: error)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        } else {
                            // process success
                            try responseString = Utils.packResponse(response)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        }
                    } catch {
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: CustomErrorMessages.errorParsingJSONResponse)
                        self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    }
                }

            });
        #endif
    }

    func unpackIsAuthorizationRequiredParams(_ command: CDVInvokedUrlCommand) throws -> (statusCode: Int32, authorizationHeaderValue: String) {
        if (command.arguments.count < 2) {
            throw CustomErrors.InvalidParameterCount(expected: 2, actual: command.arguments.count)
        }
        #if swift(>=3.0)
            guard let param0 = command.argument(at: 0) as? NSNumber else {
                throw CustomErrors.InvalidParameterType(expected: "NSNumber", actual: command.argument(at: 1) as AnyObject)
            }

            guard let param1: NSString = command.argument(at: 1) as? NSString else {
                throw CustomErrors.InvalidParameterType(expected: "String", actual: command.argument(at: 1) as AnyObject)
            }
        #else
            guard let param0 = command.argumentAtIndex(0) as? NSNumber else {
                throw CustomErrors.InvalidParameterType(expected: "NSNumber", actual: command.argumentAtIndex(1) as AnyObject)
            }

            guard let param1: NSString = command.argumentAtIndex(1) as? NSString else {
                throw CustomErrors.InvalidParameterType(expected: "String", actual: command.argumentAtIndex(1) as AnyObject)
            }
        #endif

        return (statusCode: Int32(param0.intValue), authorizationHeaderValue: param1 as String)
    }

}

File: src/android/CDVBMSAuthorizationManager.java
/*
    Copyright 2015 IBM Corp.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
        http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
package com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.core;

import android.content.Context;

import com.ibm.mobilefirstplatform.clientsdk.android.core.api.*;
import com.ibm.mobilefirstplatform.clientsdk.android.logger.api.*;

import com.ibm.mobilefirstplatform.clientsdk.android.security.api.AppIdentity;
import com.ibm.mobilefirstplatform.clientsdk.android.security.api.DeviceIdentity;
import com.ibm.mobilefirstplatform.clientsdk.android.security.api.UserIdentity;
import com.ibm.mobilefirstplatform.clientsdk.android.security.mca.api.MCAAuthorizationManager;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Map;
import java.util.List;
import java.util.HashMap;


public class CDVBMSAuthorizationManager extends CordovaPlugin {

    private static final Logger amLogger = Logger.getLogger(Logger.INTERNAL_PREFIX + "CDVBMSAuthorizationManager");

    private static final String PersistencePolicyAlways = "ALWAYS";
    private static final String PersistencePolicyNever = "NEVER";
    private static final String WWW_AUTHENTICATE_HEADER_NAME = "Www-Authenticate";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        boolean ans = true;
        if ("initialize".equals(action)) {
            this.initialize(args, callbackContext);
        } else if ("obtainAuthorizationHeader".equals(action)) {
            this.obtainAuthorizationHeader(callbackContext);
        } else if ("clearAuthorizationData".equals(action)) {
            this.clearAuthorizationData(callbackContext);
        } else if ("isAuthorizationRequired".equals(action)) {
            this.isAuthorizationRequired(args, callbackContext);
        } else if ("getCachedAuthorizationHeader".equals(action)) {
            this.getCachedAuthorizationHeader(callbackContext);
        } else if ("getAuthorizationPersistencePolicy".equals(action)) {
            this.getAuthorizationPersistencePolicy(callbackContext);
        } else if ("setAuthorizationPersistencePolicy".equals(action)) {
            this.setAuthorizationPersistencePolicy(args, callbackContext);
        } else if ("getUserIdentity".equals(action)) {
            this.getUserIdentity(callbackContext);
        } else if ("getAppIdentity".equals(action)) {
            this.getAppIdentity(callbackContext);
        } else if ("getDeviceIdentity".equals(action)) {
            this.getDeviceIdentity(callbackContext);
        }else if ("logout".equals(action)) {
            this.logout(callbackContext);
        } else {
            ans = false;
        }
        return ans;
    }

    /**
     * Use the native SDK API to initialize the authorization manager with tenantId.
     *
     * @param args            JSONArray that contains the MCA service tenantId.
     * @param callbackContext
     */
    private void initialize(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        final Context currentContext = this.cordova.getActivity();
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                String tenantId = null;
                try {
                    tenantId = args.getString(0);
                } catch (JSONException e) {
                    e.printStackTrace();
                    amLogger.error("Error in parsing the tenantId");
                    callbackContext.error("The specified tenantId cant be parse as String.");
                }
                MCAAuthorizationManager.createInstance(currentContext, tenantId);
                amLogger.debug("Authorization Manager initialize with tenantId: " + tenantId.toString());
                callbackContext.success();
            }
        });
    }

    /**
     * Use the native SDK API to invoke process for obtaining authorization header.
     * @param callbackContext Callback that will indicate whether the request succeeded or failed
     */
    private void obtainAuthorizationHeader(final CallbackContext callbackContext) throws JSONException {

        final Context currentContext = this.cordova.getActivity();

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MCAAuthorizationManager.getInstance().obtainAuthorization(currentContext, new ResponseListener() {
                    @Override
                    public void onSuccess(Response response) {
                        try {
                            PluginResult result = new PluginResult(PluginResult.Status.OK, CDVBMSRequest.packJavaResponseToJSON(response));
                            amLogger.debug("ObtainAuthorizationHeader: request successful.");
                            callbackContext.sendPluginResult(result);
                        } catch (JSONException e) {
                            callbackContext.error(e.getMessage());
                        }
                    }

                    @Override
                    public void onFailure(Response response, Throwable t, JSONObject extendedInfo) {
                        try {
                            PluginResult result = new PluginResult(PluginResult.Status.ERROR, CDVBMSRequest.packJavaResponseToJSON(response));
                            amLogger.error("Failed to send request obtainAuthorizationHeader.");
                            callbackContext.sendPluginResult(result);
                        } catch (JSONException e) {
                            callbackContext.error(e.getMessage());
                        }
                    }
                });

            }
        });

    }
    /**
     * Use the native SDK API to clear the local stored authorization data.
     * @param callbackContext
     */
    private void clearAuthorizationData(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MCAAuthorizationManager.getInstance().clearAuthorizationData();
                amLogger.debug("Authorization data cleared.");
                callbackContext.success();
            }
        });
    }

    /**
     * Use the native SDK API to check if the params came from response that requires authorization.
     *
     * @param args            JSONArray that contains the statusCode of the response, and responseAuthorizationHeader.
     * @param callbackContext Callback that will get the result,true if status is 401 or 403 and The value of the header contains 'Bearer' AND 'realm="imfAuthentication"'
     */
    private void isAuthorizationRequired(final JSONArray args, final CallbackContext callbackContext) throws JSONException {

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    int statusCode = args.getInt(0);
                    ArrayList<String> headers = new ArrayList<String>();
                    headers.add(args.getString(1));
                    Map<String, List<String>> responseAuthorizationHeader = new HashMap<String, List<String>>();
                    responseAuthorizationHeader.put(WWW_AUTHENTICATE_HEADER_NAME, headers);
                    boolean answer = MCAAuthorizationManager.getInstance().isAuthorizationRequired(statusCode, responseAuthorizationHeader);
                    amLogger.debug("isAuthorizationRequired return " + answer);
                    callbackContext.success(String.valueOf(answer));
                } catch (JSONException e) {
                    amLogger.debug("Expected non-empty string argument.");
                    callbackContext.error(e.getMessage());
                }
            }
        });
    }

    /**
     * Use the native SDK API get the locally stored authorization header or null if the value is not exist.
     *
     * @param callbackContext
     */
    private void getCachedAuthorizationHeader(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                String header = MCAAuthorizationManager.getInstance().getCachedAuthorizationHeader();
                amLogger.debug("Cached authorization header: " + header);
                callbackContext.success(header);
            }
        });
    }

    /**
     * Use the native SDK API to get the current authorization persistence policy
     *
     * @param callbackContext
     */
    private void getAuthorizationPersistencePolicy(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MCAAuthorizationManager.PersistencePolicy policy = MCAAuthorizationManager.getInstance().getAuthorizationPersistencePolicy();
                amLogger.debug("PersistencePolicy:" + policy.toString());
                callbackContext.success(policy.toString());
            }
        });
    }

    /**
     * Use the native SDK API to change the sate of the current authorization persistence policy.
     *
     * @param args            JSONArray that contains the new policy to use.
     * @param callbackContext
     */
    private void setAuthorizationPersistencePolicy(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                String newPolicy = null;
                try {
                    newPolicy = args.getString(0).toUpperCase();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                boolean success = true;
                if (newPolicy.equals(PersistencePolicyNever)) {
                    MCAAuthorizationManager.getInstance().setAuthorizationPersistencePolicy(MCAAuthorizationManager.PersistencePolicy.NEVER);
                } else if (newPolicy.equals(PersistencePolicyAlways)) {
                    MCAAuthorizationManager.getInstance().setAuthorizationPersistencePolicy(MCAAuthorizationManager.PersistencePolicy.ALWAYS);
                } else {
                    success = false;
                    amLogger.debug("Policy cann't be recognized:" + newPolicy.toString());
                    callbackContext.error("The specified persistence policy is not supported.");
                }

                if (success) {
                    amLogger.debug("PersistencePolicy set to:" + newPolicy.toString());
                    callbackContext.success();
                }
            }
        });
    }

    /**
     * Use the native SDK API to get the authorized user identity.
     *
     * @param callbackContext
     */
    private void getUserIdentity(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                UserIdentity userIdentity = MCAAuthorizationManager.getInstance().getUserIdentity();
                amLogger.debug("userIdentity: " + userIdentity.toString());
                callbackContext.success(userIdentity.toString());
            }
        });
    }

    /**
     * Use the native SDK API to get the application identity.
     *
     * @param callbackContext
     */
    private void getAppIdentity(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                AppIdentity appIdentity = MCAAuthorizationManager.getInstance().getAppIdentity();
                amLogger.debug("appIdentity: " + appIdentity.toString());
                callbackContext.success(appIdentity.toString());
            }
        });
    }

    /**
     * Use the native SDK API to get the device identity.
     *
     * @param callbackContext
     */
    private void getDeviceIdentity(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                DeviceIdentity deviceIdentity = MCAAuthorizationManager.getInstance().getDeviceIdentity();
                amLogger.debug("deviceIdentity: " + deviceIdentity.toString());
                callbackContext.success(deviceIdentity.toString());
            }
        });
    }

    /**
     * Use the native SDK API to invoke process to logout.
     * @param callbackContext Callback that will indicate whether the request succeeded or failed
     */
    private void logout(final CallbackContext callbackContext) throws JSONException {

        final Context currentContext = this.cordova.getActivity();

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {

                MCAAuthorizationManager.getInstance().logout(currentContext, new ResponseListener() {
                    @Override
                    public void onSuccess(Response response) {
                        try {
                            PluginResult result = new PluginResult(PluginResult.Status.OK, CDVBMSRequest.packJavaResponseToJSON(response));
                            amLogger.debug("Logout: request successful.");
                            callbackContext.sendPluginResult(result);
                        } catch (JSONException e) {
                            callbackContext.error(e.getMessage());
                        }
                    }

                    @Override
                    public void onFailure(Response failResponse, Throwable t, JSONObject extendedInfo) {
                        try {
                            PluginResult result = new PluginResult(PluginResult.Status.ERROR, CDVBMSRequest.packJavaResponseToJSON(failResponse));
                            amLogger.error("Failed to logout.");
                            callbackContext.sendPluginResult(result);
                        } catch (JSONException e) {
                            callbackContext.error(e.getMessage());
                        }
                    }
                });

            }
        });

    }
}